README.md
[![Build Status](https://travis-ci.org/clj-time/clj-time.svg?branch=master)](https://travis-ci.org/clj-time/clj-time)
[![Dependencies Status](https://versions.deps.co/clj-time/clj-time/status.svg)](https://versions.deps.co/clj-time/clj-time)
[![Downloads](https://versions.deps.co/clj-time/clj-time/downloads.svg)](https://versions.deps.co/clj-time/clj-time)
[![Join the chat at https://gitter.im/clj-time/clj-time](https://badges.gitter.im/clj-time/clj-time.svg)](https://gitter.im/clj-time/clj-time?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

# clj-time

A date and time library for Clojure, wrapping the [Joda Time](http://www.joda.org/joda-time/) library.

## Project Status

**The Joda Time website says:**

> Note that from Java SE 8 onwards, users are asked to migrate to java.time (JSR-310) - a core part of the JDK which replaces this project.

**If you are using Java 8 or later, consider using the built-in Java Time instead of Joda Time -- and look at [clojure.java-time](https://github.com/dm3/clojure.java-time) if you want a Clojure wrapper for that, or [cljc.java-time](https://github.com/henryw374/cljc.java-time) for a thin Clojure(Script) wrapper. See [Converting from Joda Time to java.time](http://blog.joda.org/2014/11/converting-from-joda-time-to-javatime.html) for more details about the similarities and differences between the two libraries.**

## Artifacts

`clj-time` artifacts are [released to Clojars](https://clojars.org/clj-time/clj-time).

If you are using Maven, add the following repository definition to your `pom.xml`:

``` xml
<repository>
  <id>clojars.org</id>
  <url>http://clojars.org/repo</url>
</repository>
```

### The Most Recent Release

With Leiningen:

```
[clj-time "0.15.2"]
```

With Maven:

``` xml
<dependency>
  <groupId>clj-time</groupId>
  <artifactId>clj-time</artifactId>
  <version>0.15.2</version>
</dependency>
```

## Bugs and Enhancements

Please open issues against the [official clj-time repo on Github](https://github.com/clj-time/clj-time/issues). `clj-time` is a very thin wrapper around Joda Time. That means that if Joda Time has a "peculiar behavior", it's likely to be surfaced directly in `clj-time` as well. A good example of this is `clj-time.format/unparse` which simply calls Joda Time's `.print` method -- and if the date passed in happens to be `nil`, you silently get back the _current date/time_ (many people would expect an exception!).

## Mailing List

Please ask questions on the [clj-time mailing list](http://groups.google.com/forum/#!forum/clj-time).


## Usage

### clj-time.core

The main namespace for date-time operations in the `clj-time` library is `clj-time.core`.

``` clj
(require '[clj-time.core :as t])
```

Create a DateTime instance with date-time, specifying the year, month,
day, hour, minute, second, and millisecond:


``` clj
(t/date-time 1986 10 14 4 3 27 456)
=> #<DateTime 1986-10-14T04:03:27.456Z>
```

Less-significant fields can be omitted:

``` clj
(t/date-time 1986 10 14)
=> #<DateTime 1986-10-14T00:00:00.000Z>
```

Get the current time with `now` and the start of the Unix epoch with
`epoch`.

Once you have a date-time, use accessors like `hour` and `second` to
access the corresponding fields:


```clojure
(t/hour (t/date-time 1986 10 14 22))
=> 22
```

The date-time constructor always returns times in the UTC time
zone. If you want a time with the specified fields in a different time
zone, use `from-time-zone`:


``` clj
(t/from-time-zone (t/date-time 1986 10 22) (t/time-zone-for-offset -2))
=> #<DateTime 1986-10-22T00:00:00.000-02:00>
```

If on the other hand you want a given absolute instant in time in a
different time zone, use `to-time-zone`:


``` clj
(t/to-time-zone (t/date-time 1986 10 22) (t/time-zone-for-offset -2))
=> #<DateTime 1986-10-21T22:00:00.000-02:00>
```

In addition to `time-zone-for-offset`, you can use the
`time-zone-for-id` and `default-time-zone` functions and the `utc` Var
to construct or get `DateTimeZone` instances.


If you only want a date with no time component, consider using the
`local-date` and `today` functions. These return `LocalDate` instances
that do not have time components (and thus don't suffer from
timezone-related shifting).


``` clj
(t/local-date 2013 3 20)
=> #<LocalDate 2013-03-20>
```

The functions `equal?`, `after?`, and `before?` determine the relative position
of two DateTime instances:

```clojure
(t/equal? (t/date-time 1986 10) (t/date-time 1986 10))
=> true
(t/after? (t/date-time 1986 10) (t/date-time 1986 9))
=> true
(t/before? (t/date-time 1986 9) (t/date-time 1986 10))
=> true
```

Often you will want to find a date some amount of time from a given
date. For example, to find the time 1 month and 3 weeks from a given
date-time:


``` clj
(t/plus (t/date-time 1986 10 14) (t/months 1) (t/weeks 3))
=> #<DateTime 1986-12-05T00:00:00.000Z>
```

You can perform also `minus` operations
```clj
(t/minus (t/date-time 1986 10 14 6) (t/hours 2))
=> #<DateTime 1986-10-14T04:00:00.000Z>
```


An `Interval` is used to represent the span of time between two
`DateTime` instances. Construct one using `interval`, then query them
using `within?`, `overlaps?`, and `abuts?`

```clojure
(t/within? (t/interval (t/date-time 1986) (t/date-time 1990))
              (t/date-time 1987))
=> true
```

The `in-seconds` and `in-minutes` functions can be used to describe
intervals in the corresponding temporal units:

```clojure
(t/in-minutes (t/interval (t/date-time 1986 10 2) (t/date-time 1986 10 14)))
=> 17280
```

The `overlap` function can be used to get an `Interval` representing the
overlap between two intervals:

``` clj
(t/overlap (t/interval (t/date-time 1986) (t/date-time 1990))
         (t/interval (t/date-time 1987) (t/date-time 1991)))
=> #<Interval 1987-01-01T00:00:00.000Z/1990-01-01T00:00:00.000Z>
```

`today-at` returns a moment in time at the given hour,
minute and second on the current date UTC; not the current system date:

``` clj
(t/today-at 12 00)
=> #<DateTime 2013-03-29T12:00:00.000Z>
(t/today-at 12 00 05)
=> #<DateTime 2013-03-29T12:00:05.000Z>

;; System clock says 11PM on 12/20/2016 UTC-5
(t/today-at 7 00 00)
=> #<DateTime 2016-12-21T7:00:00.000Z>
```

### clj-time.format

If you need to parse or print date-times, use `clj-time.format`:

``` clj
(require '[clj-time.format :as f])
```

Parsing and printing are controlled by formatters. You can either use
one of the built in ISO8601 formatters or define your own, e.g.:

```clojure
(def built-in-formatter (f/formatters :basic-date-time))
(def custom-formatter (f/formatter "yyyyMMdd"))
```

To see a list of available built-in formatters and an example of a
date-time printed in their format:


``` clj
(f/show-formatters)
```

Remember that `mm` is minutes, `MM` is months, `ss` is seconds and
`SS` is milliseconds. You can find a [complete list of patterns](http://www.joda.org/joda-time/key_format.html)
on the Joda Time website.

Once you have a formatter, parsing and printing are straightforward:

``` clj
(f/parse custom-formatter "20100311")
=> #<DateTime 2010-03-11T00:00:00.000Z>
```
```clojure
(f/unparse custom-formatter (t/date-time 2010 10 3))
=> "20101003"
```

To parse dates in multiple formats and format dates in just one
format, you can do this:


```clojure
(def multi-parser (f/formatter (t/default-time-zone) "YYYY-MM-dd" "YYYY/MM/dd"))

(f/unparse multi-parser (f/parse multi-parser "2012-02-01"))
=> "2012-02-01"

(f/unparse multi-parser (f/parse multi-parser "2012/02/01"))
=> "2012-02-01"
```

Note: Joda Time's `.print` method accepts a null date/time object and substitutes the current date/time, so `(f/unparse my-fmt nil)` will not throw an exception -- it will just silently return the current date/time!

### clj-time.coerce

The namespace `clj-time.coerce` contains utility functions for
coercing Joda `DateTime` instances to and from various other types:


```clojure
(require '[clj-time.coerce :as c])
```

For example, to convert a Joda `DateTime` to and from a Java `long`:

```clojure
(c/to-long (t/date-time 1998 4 25))
=> 893462400000
```
``` clj
(c/from-long 893462400000)
=> #<DateTime 1998-04-25T00:00:00.000Z>
```

And by the magic of protocols you can pass in an isoformat string and
get the unix epoch milliseconds:

```clojure
(c/to-long "2013-08-01")
=> 1375315200000
```

There are also conversions to and from `java.util.Date` (`to-date` and
`from-date`), `java.sql.Date` (`to-sql-date` and `from-sql-date`),
`java.sql.Timestamp` (`to-sql-time` and `from-sql-time`) and several
other types.

To support serialization to the ubiquitous
[EDN format](https://github.com/edn-format/edn),
`pr`, `prn` etc. will serialize Joda `DateTime` in a tagged-literal format,
that `clojure.edn/read` will deserialize.  There is a `data_readers.clj`
file, or if not loaded a `data-readers` var to use with `clojure.edn`.

```clojure
(pr-str (t/date-time 1998 4 25))
=> "#clj-time/date-time \"1998-04-25T00:00:00.000Z\""
```

```clojure
(require '[clojure.edn :as edn])
=> nil
```

```clojure
(def x (edn/read-string {:readers c/data-readers}
                        (pr-str (t/date-time 1998 4 25))))
```

```clojure
(type x)
=> org.joda.time.DateTime
```

```clj
x
=> #clj-time/date-time "1998-04-25T00:00:00.000Z"
```


### clj-time.local

The namespace `clj-time.local` contains functions for working with
local time without having to shift to/from utc, the preferred time
zone of clj-time.core.

``` clj
(require '[clj-time.local :as l])
```

Get the current local time with

``` clj
(l/local-now)
```

Get a local date-time instance retaining the time fields with

``` clj
(l/to-local-date-time obj)
```

The following all return 1986-10-14 04:03:27.246 with the local time
zone.

``` clj
(l/to-local-date-time (clj-time.core/date-time 1986 10 14 4 3 27 246))
(l/to-local-date-time "1986-10-14T04:03:27.246")
(l/to-local-date-time "1986-10-14T04:03:27.246Z")
```

The dynamic var `*local-formatters*` contains a map of local
formatters for parsing and printing. It is initialized with all the
formatters in clj-time.format localized.

to-local-date-time for strings uses `*local-formatters*` to parse.

Format an obj using a formatter in `*local-formatters*` corresponding
to the format-key passed in with

``` clj
(l/format-local-time (l/local-now) :basic-date-time)
```


### clj-time.periodic

`clj-time.periodic/periodic-seq` returns an infinite sequence of instants
separated by a time period starting with the given point in time:

``` clj
(require '[clj-time.periodic :as p])
(require '[clj-time.core :as t])

;; returns 10 instants starting with current time separated
;; by 12 hours
(take 10 (p/periodic-seq (t/now) (t/hours 12)))
```

In particular, if you ask for a sequence of instants separated by a month, you will get dates where the month increases each time (rather than being, say, 30 days apart).

### clj-time.predicates

`clj-time.predicates` comes with a set of handy predicates to
check for common conditions. For instance:

``` clj
(require '[clj-time.core :as t])
(require '[clj-time.predicates :as pr])
```
```clojure
(pr/monday? (t/date-time 1999 9 9))
=> false

(pr/january? (t/date-time 2011 1 1))
=> true

(pr/weekend? (t/date-time 2014 1 26))
=> true

(pr/weekday? (t/date-time 2014 1 26))
=> false

(pr/last-day-of-month? (t/date-time 2014 1 26))
=> false

(pr/first-day-of-month? (t/date-time 2014 1 26))
=> false
```

### clj-time.jdbc

`clj-time.jdbc` registers protocol extensions so you don’t have to use
`clj-time.coerce` yourself to coerce to and from SQL timestamps.

From the REPL:

``` clj
(require 'clj-time.jdbc)
```

In your project:

``` clj
(ns my.neat.project
  (:require [clj-time.jdbc]))

; They're registered and ready to use.
```

Now you can use `org.joda.time.DateTime` objects when "writing" to the database
in place of `java.sql.Timestamp` objects, and expect `org.joda.time.DateTime`
objects when "reading" where you would have previously expected
`java.sql.Timestamp` objects.

## Development

Running the tests:

    $ rm -f test/readme.clj && lein test-all

(assumes Leiningen 2.x)

## Documentation

The complete [API documentation](http://clj-time.github.com/clj-time/doc/index.html) is also available (codox generated).

## License

Released under the MIT License: <https://github.com/clj-time/clj-time/blob/master/MIT-LICENSE.txt>


ChangeLog.md
## Changes Between 0.15.1 and 0.15.2

* Add type hints to reduce reflection warnings for users (#268, #269, @vemv)
* Fix `overlaps?` with four arguments (#264).
* Improve tests and documentation (#258, #259, #260, @bhaskarsaraogi)

## Changes Between 0.15.0 and 0.15.1

* Add `adjust` for intervals, so they can be "moved" backwards and forwards by one or more periods (#257).

## Changes Between 0.14.5 and 0.15.0

* Add tagged literal support (in `clj-time.coerce`) (#255, @jconti). `DateTime` can be converted `to-edn` and can be printed as `#clj-time/date-time "..."`, and `data_readers.clj` provides the conversion from that back to `DateTime`.

## Changes Between 0.14.4 and 0.14.5

* Update Joda Time to 2.10 (#254, @novalis).
* Update various dev/test dependencies; clean up `spec_test.clj` to avoid Clojure refer warnings.
* Drop `oraclejdk7` from test matrix (we still support `openjdk7`); add `openjdk8`; test Clojure 1.10 on Java 8+ only (#253).
* Switch to versions.deps.co for tracking outdated dependencies (#252, @v-kolesnikov).

## Changes Between 0.14.3 and 0.14.4

* On Clojure 1.9 and later, `ReadableInstant` now extends to `Inst`, providing `inst?` and `inst-ms` on most Joda Time types (#248, #249)

## Changes Between 0.14.2 and 0.14.3

* Added deprecation notice and recommendation to move to Java Time (and [clojure.java-time](https://github.com/dm3/clojure.java-time)).
* Clarified behavior of functions when passed `nil` (due to how Joda Time handles `null` objects).
* Update to Joda Time 2.9.9.
* Drop Clojure 1.6.0 support.
* Refactor code to be more idiomatic (mostly `when` instead of `if`).

* Fully-qualify function return type hints to avoid requiring imports on use (#241).

## Changes Between 0.14.1 and 0.14.2

* Fully-qualify function return type hints to avoid requiring imports on use (#241).

## Changes Between 0.14.0 and 0.14.1

* Switches `clojure.spec` to `clojure.spec.alpha` to work with latest Clojure 1.9 builds.
* Adds `week-year` to go with `week-number-of-year` (#239, #240).
* Adds function return type hints across the board (#226).

## Changes Between 0.13.0 and 0.14.0

* Add `from-epoch`.
* Clarify `today-at` is UTC.
* Add optional `clj-time.spec` namespace (requires Clojure 1.9 Alpha 17).
* Add `clj-time.types` namespace with type-based predicates.

## Changes Between 0.12.2 and 0.13.0

* Update Joda Time to 2.9.7 (#223).

## Changes Between 0.12.1 and 0.12.2

* More reflection warnings removed (#221).

## Changes Between 0.12.0 and 0.12.1

* Updated Joda Time (to 2.9.4) and several testing dependencies as well.
* `floor` now retains timezone information (#204).
* Reflection warning removed from `with-time-at-start-of-day` (#219).

## Changes Between 0.11.0 and 0.12.0 (June 6th, 2016)

### Joda Time 2.9.3

[Joda Time](http://www.joda.org/joda-time/) has been [upgraded to `2.9.3`](http://www.joda.org/joda-time/upgradeto293.html).

### java.jdbc coercions

`java.sql.Date` and `java.sql.Time` are now coerced automatically if you `require` `clj-time.jdbc` (previously only `java.sql.Timestamp` was coerced).

### Formatters

The `formatter` function now accepts keywords and formatter objects, as well as string formats, so you can easily select a standard format (with a keyword) or provide a specific formatter object.

### Clojure 1.8 by Default

The library now depends on Clojure `1.8.0` and is tested against `1.6.0`, `1.7.0`, and `1.9.0-master-SNAPSHOT`.

### Enhancements

* `first-day-of-the-month?` and `last-day-of-the-month?` -- aliases to improve consistency (so it's `the-month` everywhere).
* `min-date` and `max-date` -- to return the minimum or maximum of two or more dates.
* `nth-day-of-the-month` -- return the Nth day of the year/month, or date/time.
* `nth-day-of-the-month?` -- return true if a given date/time is on the Nth day of its month.
* `with-time-at-start-of-day` -- return the time at the start of the day, including timezone and DST support.

### Deprecations

* `today-at-midnight` -- deprecated in favor of `with-time-at-start-of-day`.

### Bug Fixes

[Issue 185](https://github.com/clj-time/clj-time/issues/185) is fixed by explicitly tying the `:rfc822` formatter to the US locale.

The examples in the README are now automatically tested (by midje-readme) so they are more likely to be valid.

## Changes Between 0.10.0 and 0.11.0

### Joda Time 2.8

[Joda Time](http://www.joda.org/joda-time/) has been [upgraded to `2.8.2`](http://www.joda.org/joda-time/upgradeto282.html).

### Clojure 1.7 by Default

The library now depends on Clojure `1.7.0`.



## Changes Between 0.9.0 and 0.10.0

### clj-time.instant

`clj-time.instant` is a new namespace that, when loaded, makes it possible
for the Clojure reader to serialize Joda Time data types.

### Joda Time 2.7

[Joda Time](http://www.joda.org/joda-time/) has been [upgraded to `2.7`](http://www.joda.org/joda-time/upgradeto270.html).

### `equal?`

A new predicate `equal?` has been added, which can be used like `before?` and
`after?` to assert two instants represent identical points in time regardless of
differences in time zone.

Contributed by [@gws](https://github.com/gws).

### `clj-time.jdbc`

A new namespace `clj-time.jdbc` has been added which extends protocols in
clojure.java.jdbc to allow for easy coercion to and from java.sql.Timestamp
driven by `clj-time.coerce`.

Contributed by [@gws](https://github.com/gws).

## Changes Between 0.8.0 and 0.9.0

### with-default-year

`clj-time.format/with-default-year` is a new function that accepts a formatter
and returns a formatter that uses provided year by default.

Contributed by lummax.

### InTimeUnitProtocol

`clj-time.core/InTimeUnitProtocol` is a new protocol that provides convenient
functions for time unit conversion:

``` clojure
(defprotocol InTimeUnitProtocol
  "Interface for in-<time unit> functions"
  (in-millis  [this] "Return the time in milliseconds.")
  (in-seconds [this] "Return the time in seconds.")

test/clj_time/jdbc_test.clj
(ns clj-time.jdbc-test
  (:require [clojure.java.jdbc :as jdbc]
            [clojure.test :refer :all]
            [clj-time.jdbc]))

(deftest test-extends-IResultSetReadColumn
  (is (extends? jdbc/IResultSetReadColumn java.sql.Timestamp)))

(deftest test-extends-ISQLValue
  (is (extends? jdbc/ISQLValue org.joda.time.DateTime)))


test/clj_time/local_test.clj
(ns clj-time.local-test
  (:require [clojure.test :refer :all]
            [clj-time [core :as time] [format :as fmt] [local :refer :all]]
            [clj-time.core-test :refer [when-available when-not-available]])
  (:import [org.joda.time.format ISODateTimeFormat]
           java.util.Date java.sql.Timestamp))

(deftest test-now
  (is (= (time/from-time-zone (time/date-time 2010 1 1) (time/default-time-zone))
         (time/do-at (time/from-time-zone (time/date-time 2010 1 1) (time/default-time-zone)) 
                     (local-now)))))

(deftest test-to-local-date-time
  (is (nil? (to-local-date-time nil)))
  (is (nil? (to-local-date-time "")))
  (is (nil? (to-local-date-time "x")))
  (is (= (time/from-time-zone (time/date-time 1998 4 25) (time/default-time-zone)) (to-local-date-time (time/date-time 1998 4 25))))
  (is (= (time/to-time-zone (time/date-time 1998 4 25) (time/default-time-zone)) (to-local-date-time (Date. 893462400000))))
  (is (= (time/to-time-zone (time/date-time 1998 4 25) (time/default-time-zone)) (to-local-date-time (java.sql.Date. 893462400000))))
  (is (= (time/from-time-zone (time/date-time 1970 1 1) (time/default-time-zone)) (to-local-date-time 0)))
  (is (= (time/from-time-zone (time/date-time 1998 4 25) (time/default-time-zone)) (to-local-date-time 893462400000)))
  (is (= (time/to-time-zone (time/date-time 1998 4 25) (time/default-time-zone)) (to-local-date-time (Timestamp. 893462400000))))
  (is (= (time/from-time-zone (time/date-time 1998 4 25) (time/default-time-zone)) (to-local-date-time "1998-04-25T00:00:00.000"))))

(deftest test-format-local-time
  (is (nil? (format-local-time nil :basic-date-time)))
  (is (nil? (format-local-time (local-now) ::bad-formatter)))
  (is (nil? (format-local-time "bad-time" :basic-date-time)))
  (is (nil? (format-local-time "bad-time" ::bad-formatter)))
  (is (= (fmt/unparse (ISODateTimeFormat/basicDateTime) (time/from-time-zone (time/date-time 1998 4 25) (time/default-time-zone)))
         (format-local-time (time/from-time-zone (time/date-time 1998 4 25) (time/default-time-zone)) :basic-date-time)))
  (is (= (fmt/unparse (ISODateTimeFormat/basicDateTime) (time/from-time-zone (time/date-time 1998 4 25) (time/default-time-zone)))
         (format-local-time (time/date-time 1998 4 25) :basic-date-time)))
  (is (= (fmt/unparse (ISODateTimeFormat/basicDateTime) (time/to-time-zone (time/date-time 1998 4 25) (time/default-time-zone)))
         (format-local-time (Date. 893462400000) :basic-date-time)))
  (is (= (fmt/unparse (ISODateTimeFormat/basicDateTime) (time/to-time-zone (time/date-time 1998 4 25) (time/default-time-zone)))
         (format-local-time (java.sql.Date. 893462400000) :basic-date-time)))
  (is (= (fmt/unparse (ISODateTimeFormat/basicDateTime) (time/from-time-zone (time/date-time 1970 1 1) (time/default-time-zone)))
         (format-local-time 0 :basic-date-time)))
  (is (= (fmt/unparse (ISODateTimeFormat/basicDateTime) (time/from-time-zone (time/date-time 1998 4 25) (time/default-time-zone)))
         (format-local-time 893462400000 :basic-date-time)))
  (is (= (fmt/unparse (ISODateTimeFormat/basicDateTime) (time/to-time-zone (time/date-time 1998 4 25) (time/default-time-zone)))
         (format-local-time (Timestamp. 893462400000) :basic-date-time)))
  (is (= (fmt/unparse (ISODateTimeFormat/basicDateTime) (time/from-time-zone (time/date-time 1998 4 25) (time/default-time-zone)))
         (format-local-time "1998-04-25T00:00:00.000" :basic-date-time))))

(deftest test-local-formatters
  (letfn [(time-zone-fn []  (time/time-zone-for-offset -7))
          (asserts []
            (let [formatters {:mmddyyyy-slash (fmt/formatter "MM/dd/yyyy" (time/default-time-zone))
                              :mmddyyyy-hhmmss-slash (fmt/formatter "MM/dd/yyyy HH:mm:ss" (time/default-time-zone))}]
              (binding [*local-formatters* formatters]
                (is (= (time/time-zone-for-offset -7) (time/default-time-zone)))
                (is (= "04/25/1998" (format-local-time (time/date-time 1998 4 25 11 59 1) :mmddyyyy-slash)))
                (is (= "04/25/1998 11:59:01" (format-local-time (time/date-time 1998 4 25 11 59 1) :mmddyyyy-hhmmss-slash)))
                (is (= (time/from-time-zone (time/date-time 1998 4 25 11 59 1) (time/default-time-zone))
                       (to-local-date-time "04/25/1998 11:59:01")))
                (is (= (time/default-time-zone) (.getZone (to-local-date-time "04/25/1998 11:59:01")))))))]
    (when-available
     with-redefs
     (with-redefs [time/default-time-zone time-zone-fn]
       (asserts)))
    (when-not-available
     with-redefs
     (binding [time/default-time-zone time-zone-fn]
       (asserts)))))


src/clj_time/local.clj
(ns clj-time.local
  "Functions for working with local time without having to shift
   to/from utc, the preferred time zone of clj-time.core.

   Get the current local time with (local-now).

   (to-local-date-time obj) returns a local date-time instance
   retaining the time fields.

  The following all return 1986-10-14 04:03:27.246 with the
  local time zone.

  (to-local-date-time (clj-time.core/date-time 1986 10 14 4 3 27 246))
  (to-local-date-time \"1986-10-14T04:03:27.246\")
  (to-local-date-time \"1986-10-14T04:03:27.246Z\")

  The dynamic var *local-formatters* contains a map of local formatters
  for parsing and printing. It is initialized with all the formatters in
  clj-time.format localized.

  to-local-date-time for strings uses *local-formatters* to parse.

  (format-local-time (local-now) :basic-date-time) formats an obj using
  a formatter in *local-formatters* corresponding to the  format-key
  passed in.
"
  (:require [clj-time.core :as time]
            [clj-time.coerce :as coerce]
            [clj-time.format :as fmt])
  (:import (org.joda.time DateTime DateTimeZone)
           (org.joda.time.format DateTimeFormatter)))

(def ^{:doc "Map of local formatters for parsing and printing." :dynamic true}
  *local-formatters*
  (into {} (map
            (fn [[k ^DateTimeFormatter f]] [k (.withZone f ^DateTimeZone (time/default-time-zone))])
            fmt/formatters)))

(defn local-now
  "Returns a DateTime for the current instant in the default time zone."
  []
  (DateTime/now ^DateTimeZone (time/default-time-zone)))

(defprotocol ILocalCoerce
  (to-local-date-time [obj] "convert `obj` to a local Joda DateTime instance retaining time fields."))

(defn- as-local-date-time-from-time-zone
  "Coerce to date-time in the default time zone retaining time fields."
  [obj]
  (-> obj coerce/to-date-time (time/from-time-zone (time/default-time-zone))))

(defn- as-local-date-time-to-time-zone
  "Coerce to date-time in the default time zone."
  [obj]
  (-> obj coerce/to-date-time (time/to-time-zone (time/default-time-zone))))

(defn- from-local-string
  "Return local DateTime instance from string using
   formatters in *local-formatters*, returning first
   which parses."
  [s]
  (first
   (for [f (vals *local-formatters*)
         :let [d (try (fmt/parse f s) (catch Exception _ nil))]
         :when d] d)))

(extend-protocol ILocalCoerce
  nil
  (to-local-date-time [_]
    nil)

  java.util.Date
  (to-local-date-time [date]
    (as-local-date-time-to-time-zone date))

  java.sql.Date
  (to-local-date-time [sql-date]
    (as-local-date-time-to-time-zone sql-date))

  DateTime
  (to-local-date-time [date-time]
    (as-local-date-time-from-time-zone date-time))

  Integer
  (to-local-date-time [integer]
    (as-local-date-time-from-time-zone (long integer)))

  Long
  (to-local-date-time [long]
    (as-local-date-time-from-time-zone long))

  String
  (to-local-date-time [string]
    (from-local-string string))

  java.sql.Timestamp
  (to-local-date-time [timestamp]
    (as-local-date-time-to-time-zone timestamp)))

(defn format-local-time
  "Format obj as local time using the local formatter corresponding
   to format-key."
  [obj format-key]
  (when-let [dt (to-local-date-time obj)]
    (when-let [fmt (format-key *local-formatters*)]
      (fmt/unparse fmt dt))))


src/clj_time/instant.clj
(ns clj-time.instant
    "An optional convenience namespaces that allows key JodaTime types
    to be transparently serialized with the Clojure reader (via instant literals)."
    )


(defmethod print-dup java.util.Date
  [^java.util.Date d ^java.io.Writer out]
  (.write out
          (str "#="
               `(java.util.Date. ~(.getYear d)
                                 ~(.getMonth d)
                                 ~(.getDate d)
                                 ~(.getHours d)
                                 ~(.getMinutes d)
                                 ~(.getSeconds d)))))


(defmethod print-dup org.joda.time.base.AbstractInstant
  [^org.joda.time.base.AbstractInstant d out]
  (print-dup (.toDate d) out))


(defmethod print-dup org.joda.time.base.AbstractPartial
  [^org.joda.time.base.AbstractPartial d out]
  (print-dup (.toDate d) out))


src/clj_time/periodic.clj
(ns clj-time.periodic
  (:require [clj-time.core :as ct])
  (:import [org.joda.time DateTime ReadablePeriod Period]))

(defn periodic-seq
  "Returns a sequence of date-time values growing over specific period.
  The 2 argument function takes as input the starting value and the growing value,
  returning a lazy infinite sequence.
  The 3 argument function takes as input the starting value, the upper bound value,
  and the growing value, return a lazy sequence."
  ([^DateTime start ^ReadablePeriod period-like]
   (let [^Period period (.toPeriod period-like)]
     (map (fn [i]
            (ct/plus start (.multipliedBy period i)))
          (iterate inc 0))))
  ([^DateTime start ^DateTime end ^ReadablePeriod period-like]
   (let [^Period period (.toPeriod period-like)]
     (->> (iterate inc 0)
          (map (fn [i]
                 (ct/plus start (.multipliedBy period i))))
          (take-while (fn [^DateTime next]
                        (ct/before? next end)))))))


src/clj_time/predicates.clj
(ns clj-time.predicates
  "Predicate functions to ask basic questions about a date

  Was it Monday?
  (monday? (clj-time.core/date-time 1999 9 9))

  Is it January?
  (january? (clj-time.core/date-time 2011 1 1))"
  (:require [clj-time.core :as time]
            [clj-time.coerce :as coerce]))

;; days of the week
(defn monday? [date-time]
  (= (time/day-of-week date-time) 1))

(defn tuesday? [date-time]
  (= (time/day-of-week date-time) 2))

(defn wednesday? [date-time]
  (= (time/day-of-week date-time) 3))

(defn thursday? [date-time]
  (= (time/day-of-week date-time) 4))

(defn friday? [date-time]
  (= (time/day-of-week date-time) 5))

(defn saturday? [date-time]
  (= (time/day-of-week date-time) 6))

(defn sunday? [date-time]
  (= (time/day-of-week date-time) 7))

;; weekend / weekday checks
(defn weekend? [date-time] 
  (or (saturday? date-time) (sunday? date-time)))

(defn weekday? [date-time]
  (not (weekend? date-time)))

;; months of the year
(defn january? [date-time]
  (= (time/month date-time) 1))

(defn february? [date-time]
  (= (time/month date-time) 2))

(defn march? [date-time]
  (= (time/month date-time) 3))

(defn april? [date-time]
  (= (time/month date-time) 4))

(defn may? [date-time]
  (= (time/month date-time) 5))

(defn june? [date-time]
  (= (time/month date-time) 6))

(defn july? [date-time]
  (= (time/month date-time) 7))

(defn august? [date-time]
  (= (time/month date-time) 8))

(defn september? [date-time]
  (= (time/month date-time) 9))

(defn october? [date-time]
  (= (time/month date-time) 10))

(defn november? [date-time]
  (= (time/month date-time) 11))

(defn december? [date-time]
  (= (time/month date-time) 12))


;;First and last day of month checks
(defn last-day-of-month? [date-time]
  (= (time/last-day-of-the-month date-time) date-time))

(defn first-day-of-month? [date-time]
  (= (time/first-day-of-the-month date-time) date-time))

(defn same-date?
  "Compares two date times to see if they are the same date"
  [this-date-time that-date-time]
  (= (coerce/to-local-date this-date-time) (coerce/to-local-date that-date-time)))


project.clj
(def java7? (.startsWith (System/getProperty "java.version") "1.7"))

(defproject clj-time/clj-time "0.15.3-SNAPSHOT"
  :description "A date and time library for Clojure, wrapping Joda Time."
  :url "https://github.com/clj-time/clj-time"
  :mailing-list {:name "clj-time mailing list"
                 :archive "https://groups.google.com/forum/?fromgroups#!forum/clj-time"
                 :post "clj-time@googlegroups.com"}
  :license {:name "MIT License"
            :url "http://www.opensource.org/licenses/mit-license.php"
            :distribution :repo}
  :dependencies [[joda-time "2.10"]
                 [org.clojure/clojure "1.10.0" :scope "provided"]]
  :min-lein-version "2.0.0"
  :global-vars {*warn-on-reflection* true}
  :profiles {:dev {:dependencies [[org.clojure/java.jdbc "0.7.9"]]
                   :plugins [[codox "0.10.6"]]}
             :midje {:dependencies [[midje "1.9.8"]]
                     :plugins      [[lein-midje "3.2.1"]
                                    [midje-readme "1.0.9"]]
                     :midje-readme {:require "[clj-time.core :as t] [clj-time.predicates :as pr] [clj-time.format :as f] [clj-time.coerce :as c]"}}
             :1.7    {:dependencies [[org.clojure/clojure "1.7.0"]]}
             :1.8    {:dependencies [[org.clojure/clojure "1.8.0"]]}
             :1.9    {:dependencies [[org.clojure/clojure "1.9.0"]
                                     [org.clojure/test.check "0.10.0-alpha4"]]
                      :test-paths ["test" "test_spec"]}
             :1.10   {:dependencies [[org.clojure/clojure "1.10.0"]
                                     [org.clojure/test.check "0.10.0-alpha4"]]
                      :test-paths ["test" "test_spec"]}
             :master {:repositories [["snapshots" "https://oss.sonatype.org/content/repositories/snapshots/"]]
                      :dependencies [[org.clojure/clojure "1.11.0-master-SNAPSHOT"]
                                     [org.clojure/test.check "0.10.0-alpha4"]]
                      :test-paths ["test" "test_spec"]}}

  :aliases {"test-all" ["with-profile"
                        ~(str (when-not java7?
                                ;; 1.10+ requires Java 8+
                                (str "dev,master,midje:" ; 1.11 with spec
                                     "dev,1.10,midje:" ; 1.10 with spec
                                     "dev,default,midje:")) ; 1.10 without spec
                              "dev,1.9,midje:" ; 1.9 with spec
                              "dev,1.8,midje:" ; 1.8 is supported too
                              "dev,1.7,midje") ; 1.7 is earliest we support
                        "test"]})
